source/common/router/rds_impl.cc

```diff
@@ -251,37 +251,38 @@ RouteConfigProviderManagerImpl::RouteConfigProviderManagerImpl(Server::Admin& ad
   RELEASE_ASSERT(config_tracker_entry_, "");
 }
 
-Router::RouteConfigProviderPtr RouteConfigProviderManagerImpl::createRdsRouteConfigProvider(
+Router::RouteConfigProviderSharedPtr RouteConfigProviderManagerImpl::createRdsRouteConfigProvider(
     const envoy::config::filter::network::http_connection_manager::v2::Rds& rds,
     Server::Configuration::FactoryContext& factory_context, const std::string& stat_prefix,
     Init::Manager& init_manager) {
   // RdsRouteConfigSubscriptions are unique based on their serialized RDS config.
   const uint64_t manager_identifier = MessageUtil::hash(rds);
-  auto& server_factory_context = factory_context.getServerFactoryContext();
+  auto it = dynamic_route_config_providers_.find(manager_identifier);
 
-  RdsRouteConfigSubscriptionSharedPtr subscription;
-
-  auto it = route_config_subscriptions_.find(manager_identifier);
-  if (it == route_config_subscriptions_.end()) {
+  if (it == dynamic_route_config_providers_.end()) {
     // std::make_shared does not work for classes with private constructors. There are ways
     // around it. However, since this is not a performance critical path we err on the side
     // of simplicity.
-    subscription.reset(new RdsRouteConfigSubscription(
-        rds, manager_identifier, server_factory_context, factory_context.messageValidationVisitor(),
-        init_manager, stat_prefix, *this));
+    RdsRouteConfigSubscriptionSharedPtr subscription(new RdsRouteConfigSubscription(
+        rds, manager_identifier, factory_context.getServerFactoryContext(),
+        factory_context.messageValidationVisitor(), factory_context.initManager(), stat_prefix,
+        *this));
     init_manager.add(subscription->init_target_);
-    route_config_subscriptions_.insert({manager_identifier, subscription});
+    std::shared_ptr<RdsRouteConfigProviderImpl> new_provider{
+        new RdsRouteConfigProviderImpl(std::move(subscription), factory_context)};
+    dynamic_route_config_providers_.insert(
+        {manager_identifier, std::weak_ptr<RdsRouteConfigProviderImpl>(new_provider)});
```

Comment:
Do we have to hold a strong reference to ensure this doesn't destruct on one of the worker threads?

Reply:
Subscription holds a strong reference of Provider here: https://github.com/envoyproxy/envoy/blob/fb628ce25441d5f7d424fee17f9d773c335bc152/source/common/router/rds_impl.cc#L218

And map look up will only see two possibility:
1) no such key in map, no matter worker threads hold a strong reference or not. 
2) map has such key. It means the `subscription` is not destructed. That subscription hold a strong ref to the provider.

Both possibilities are under the assumption that 
`subscription exist iff key exists in this dynamic_route_config_providers_ map`
I think this assumption is correct because both the CRUD on map and construct/destruct on subscription occurs on main thread.
See Xin's comments 4 lines below.

Reply:
My above comment could be wrong. I will add some ASSERT